src/lib/log.c

```diff
@@ -19,19 +19,94 @@
 
 #define _GNU_SOURCE
 
+#include <ctype.h>
 #include <stdarg.h>
+#include <stdlib.h>
+#include <string.h>
 #include <time.h>
 
 #include "log.h"
 #define DEBUG_HEADER_SIZE 50
 
 static int cur_log_level = LOG_INFO;
+static int json_format = 0;
 
 void set_log_level(int log_level)
 {
 	cur_log_level = log_level;
 }
 
+void set_json_format(void)
+{
+	json_format = 1;
+}
+
+void json_start(const char *op)
+{
+	if (json_format) {
+		fprintf(stderr, "[\n");
+		fflush(stderr);
+		fprintf(stderr, "{ \"%s\" : \"start\" },\n", op);
+		fflush(stderr);
+	}
+}
+
+void json_end(const char *op)
+{
+	if (json_format) {
+		fprintf(stderr, "{ \"%s\" : \"end\" }\n", op);
+		fflush(stderr);
+		fprintf(stderr, "]\n");
+		fflush(stderr);
+	}
+}
+
+void format_to_json(const char *type, const char *msg, va_list args_list)
+{
+	const int MAX_TYPE_LENGTH = 8;
+	char *full_msg;
+	char ltype[MAX_TYPE_LENGTH];
+
+	/* make sure the type is all lower case */
+	if (type) {
+		for (int i = 0; type[i]; i++) {
+			ltype[i] = tolower(type[i]);
+		}
+		ltype[strlen(type)] = '\0';
+	} else {
+		/* if no type was provided we default to info */
+		strcpy(ltype, "info");
+	}
+
+	/* build the full message based on all the arguments */
+	if (vasprintf(&full_msg, msg, args_list) < 0) {
+		abort();
+	}
+
+	if (strcmp(full_msg, "") != 0) {
+		int j = 0;
+		for (unsigned int i = 0; i <= strlen(full_msg); i++) {
+			/* if the message has double quotes replace them with single quotes
+			* since double quotes are not supported in JSON */
+			if (full_msg[i] == '"') {
```

Comment:
actually " are supported, but we need to escape them (https://www.json.org has a really good chart for json format). I'm ok on changing that to ' though.

Reply:
and '\' isn't supported too.

Reply:
> and '' isn't supported too.

I didn't get this

Reply:
Probably I removed the character by accident. I was trying to say that \ isn't a valid json character too because it's used as a escape char. We need to replace all \ to \\ (or just remove \ because probably we aren't even using it).